app/scripts/controllers/selected-network-controller.ts

```diff
@@ -0,0 +1,154 @@
+  domains: Record<Domain, NetworkClientId>;
+  // Feature flag to start returning networkClientId based on the domain.
+  // when the flag is false, the 'metamask' domain will always be used.
+  // defaults to false
+  perDomainNetwork: boolean;
+};
+
+export type GetSelectedNetworkState = {
+  type: `${typeof controllerName}:getState`;
+  handler: () => SelectedNetworkControllerState;
+};
+
+export type GetSelectedNetworkStateChange = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [SelectedNetworkControllerState, Patch[]];
+};
+
+export type GetNetworkClientIdForDomain = {
+  type: `SelectedNetworkController:getNetworkClientIdForDomain`;
+  handler: (domain: string) => NetworkClientId;
+};
+
+export type SetNetworkClientIdForDomain = {
+  type: `SelectedNetworkController:setNetworkClientIdForDomain`;
+  handler: (domain: string, NetworkClientId: NetworkClientId) => void;
+};
+
+export type SelectedNetworkControllerActions =
+  | GetSelectedNetworkState
+  | GetNetworkClientIdForDomain
+  | SetNetworkClientIdForDomain;
+
+export type SelectedNetworkControllerEvents = GetSelectedNetworkStateChange;
+
+export type SelectedNetworkControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  SelectedNetworkControllerActions,
+  NetworkControllerStateChangeEvent | SelectedNetworkControllerEvents,
+  never,
+  NetworkControllerStateChangeEvent['type']
+>;
+
+export type SelectedNetworkControllerOptions = {
+  messenger: SelectedNetworkControllerMessenger;
+};
+
+/**
+ * Controller for getting and setting the network for a particular domain.
+ */
+export default class SelectedNetworkController extends BaseControllerV2<
+  typeof controllerName,
+  SelectedNetworkControllerState,
+  SelectedNetworkControllerMessenger
+> {
+  /**
+   * Construct a SelectedNetworkController controller.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The restricted controller messenger for the EncryptionPublicKey controller.
+   */
+  constructor({ messenger }: SelectedNetworkControllerOptions) {
+    super({
+      name: controllerName,
+      metadata: stateMetadata,
+      messenger,
+      state: getDefaultState(),
+    });
+    this.registerMessageHandlers();
+  }
+
+  private registerMessageHandlers(): void {
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:getNetworkClientIdForDomain` as const,
+      this.getNetworkClientIdForDomain.bind(this),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:setNetworkClientIdForDomain` as const,
+      this.setNetworkClientIdForDomain.bind(this),
+    );
+
+    // subscribe to networkController statechange:: selectedNetworkClientId changed
+    // update the value for the domain 'metamask'
+    this.messagingSystem.subscribe(
+      'NetworkController:stateChange',
+      (state: NetworkState, patch: Patch[]) => {
+        const isChangingNetwork = patch.find(
```

Comment:
nit, feel free to ignore. `patch.some()` since we don't actually care about the result itself, just that it exists?